# Image Classification
[How it works](https://apple.github.io/turicreate/docs/userguide/image_classifier/how-it-works.html)
## Loading Data

In [1]:
import turicreate as tc

In [2]:
data = tc.image_analysis.load_images('kagglecatsanddogs_3367a/PetImages', with_path=True)

In [4]:
data

Columns:
	path	str
	image	Image

Rows: 24769

Data:
+-------------------------------+------------------------+
|              path             |         image          |
+-------------------------------+------------------------+
| /Users/st21235/Project/Jup... | Height: 375 Width: 500 |
| /Users/st21235/Project/Jup... | Height: 281 Width: 300 |
| /Users/st21235/Project/Jup... | Height: 500 Width: 489 |
| /Users/st21235/Project/Jup... | Height: 500 Width: 403 |
| /Users/st21235/Project/Jup... | Height: 150 Width: 150 |
| /Users/st21235/Project/Jup... | Height: 360 Width: 431 |
| /Users/st21235/Project/Jup... | Height: 375 Width: 500 |
| /Users/st21235/Project/Jup... | Height: 471 Width: 500 |
| /Users/st21235/Project/Jup... | Height: 375 Width: 500 |
| /Users/st21235/Project/Jup... | Height: 240 Width: 320 |
+-------------------------------+------------------------+
[24769 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [3]:
data['label'] = data['path'].apply(lambda path: 'dog' if '/Dog' in path else 'cat')

In [17]:
data['label']

dtype: str
Rows: 24769
['cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', ... ]

**Save the data for future use**

In [4]:
data.save('cats-dogs.sframe')

**Explore interactively**

In [4]:
data.explore()

Materializing SFrame...

Done.

# Predict the picture

**Load the data from SFrame**

In [3]:
data = tc.SFrame('cats-dogs.sframe')

In [4]:
train_data, test_data = data.random_split(0.8)

**Create a model**

In [5]:
model = tc.image_classifier.create(train_data, target='label')

Download completed: /var/folders/m4/8rlv2k612l9g_hd_gmmlbg5m0000gp/T/model_cache/resnet-50-symbol.json
Download completed: /var/folders/m4/8rlv2k612l9g_hd_gmmlbg5m0000gp/T/model_cache/resnet-50-0000.params


Analyzing and extracting image features.

+------------------+--------------+------------------+

| Images Processed | Elapsed Time | Percent Complete |

+------------------+--------------+------------------+

| 0                | 1.33s        | 0%               |

| 64               | 6.36s        | 0.25%            |

| 128              | 7.60s        | 0.5%             |

| 192              | 8.85s        | 0.75%            |

| 256              | 10.09s       | 1.25%            |

| 320              | 11.34s       | 1.5%             |

| 384              | 12.59s       | 1.75%            |

| 448              | 13.83s       | 2.25%            |

| 512              | 15.08s       | 2.5%             |

| 576              | 16.36s       | 2.75%            |

| 640              | 17.63s       | 3%               |

| 704              | 18.90s       | 3.5%             |

| 768              | 20.17s       | 3.75%            |

| 832              | 21.47s       | 4%               |

| 896              | 22.72s       | 4.5%             |

| 960              | 23.97s       | 4.75%            |

| 1024             | 25.21s       | 5%               |

| 1088             | 26.49s       | 5.5%             |

| 1152             | 27.84s       | 5.75%            |

| 1216             | 29.14s       | 6%               |

| 1280             | 30.42s       | 6.25%            |

| 1344             | 31.71s       | 6.75%            |

| 1408             | 32.95s       | 7%               |

| 1472             | 34.27s       | 7.25%            |

| 1536             | 35.58s       | 7.75%            |

| 1600             | 36.84s       | 8%               |

| 1664             | 38.08s       | 8.25%            |

| 1728             | 39.33s       | 8.5%             |

| 1792             | 40.58s       | 9%               |

| 1856             | 41.84s       | 9.25%            |

| 1920             | 43.15s       | 9.5%             |

| 1984             | 44.42s       | 10%              |

| 2048             | 45.68s       | 10.25%           |

| 2112             | 46.92s       | 10.5%            |

| 2176             | 48.16s       | 11%              |

| 2240             | 49.44s       | 11.25%           |

| 2304             | 50.77s       | 11.5%            |

| 2368             | 52.04s       | 11.75%           |

| 2432             | 53.32s       | 12.25%           |

| 2496             | 54.59s       | 12.5%            |

| 2624             | 57.11s       | 13.25%           |

| 2688             | 58.42s       | 13.5%            |

| 2752             | 59.79s       | 13.75%           |

| 2816             | 1m 1s        | 14%              |

| 2880             | 1m 2s        | 14.5%            |

| 2944             | 1m 3s        | 14.75%           |

| 3008             | 1m 4s        | 15%              |

| 3072             | 1m 6s        | 15.5%            |

| 3136             | 1m 7s        | 15.75%           |

| 3200             | 1m 8s        | 16%              |

| 3264             | 1m 9s        | 16.5%            |

| 3328             | 1m 11s       | 16.75%           |

| 3392             | 1m 12s       | 17%              |

| 3456             | 1m 13s       | 17.25%           |

| 3520             | 1m 14s       | 17.75%           |

| 3584             | 1m 15s       | 18%              |

| 3648             | 1m 17s       | 18.25%           |

| 3712             | 1m 18s       | 18.75%           |

| 3776             | 1m 19s       | 19%              |

| 3840             | 1m 20s       | 19.25%           |

| 3904             | 1m 22s       | 19.5%            |

| 3968             | 1m 23s       | 20%              |

| 4032             | 1m 24s       | 20.25%           |

| 4096             | 1m 25s       | 20.5%            |

| 4160             | 1m 27s       | 21%              |

| 4224             | 1m 28s       | 21.25%           |

| 4288             | 1m 29s       | 21.5%            |

| 4352             | 1m 30s       | 22%              |

| 4416             | 1m 32s       | 22.25%           |

| 4480             | 1m 33s       | 22.5%            |

| 4544             | 1m 34s       | 22.75%           |

| 4608             | 1m 35s       | 23.25%           |

| 4672             | 1m 37s       | 23.5%            |

| 4736             | 1m 38s       | 23.75%           |

| 4800             | 1m 39s       | 24.25%           |

| 4864             | 1m 40s       | 24.5%            |

| 4928             | 1m 41s       | 24.75%           |

| 4992             | 1m 43s       | 25%              |

| 5056             | 1m 44s       | 25.5%            |

| 5120             | 1m 45s       | 25.75%           |

| 5184             | 1m 46s       | 26%              |

| 5248             | 1m 48s       | 26.5%            |

| 5312             | 1m 49s       | 26.75%           |

| 5376             | 1m 50s       | 27%              |

| 5440             | 1m 51s       | 27.5%            |

| 5504             | 1m 53s       | 27.75%           |

| 5568             | 1m 54s       | 28%              |

| 5632             | 1m 55s       | 28.25%           |

| 5696             | 1m 56s       | 28.75%           |

| 5760             | 1m 58s       | 29%              |

| 5824             | 1m 59s       | 29.25%           |

| 5888             | 2m 0s        | 29.75%           |

| 5952             | 2m 1s        | 30%              |

| 6016             | 2m 3s        | 30.25%           |

| 6080             | 2m 4s        | 30.5%            |

| 6144             | 2m 5s        | 31%              |

| 6208             | 2m 6s        | 31.25%           |

| 6272             | 2m 8s        | 31.5%            |

| 6336             | 2m 9s        | 32%              |

| 6400             | 2m 10s       | 32.25%           |

| 6464             | 2m 11s       | 32.5%            |

| 6528             | 2m 13s       | 33%              |

| 6592             | 2m 14s       | 33.25%           |

| 6656             | 2m 15s       | 33.5%            |

| 6720             | 2m 16s       | 33.75%           |

| 6784             | 2m 17s       | 34.25%           |

| 6848             | 2m 19s       | 34.5%            |

| 6912             | 2m 20s       | 34.75%           |

| 6976             | 2m 21s       | 35.25%           |

| 7040             | 2m 22s       | 35.5%            |

| 7104             | 2m 24s       | 35.75%           |

| 7168             | 2m 25s       | 36%              |

| 7232             | 2m 26s       | 36.5%            |

| 7296             | 2m 28s       | 36.75%           |

| 7360             | 2m 29s       | 37%              |

| 7424             | 2m 30s       | 37.5%            |

| 7488             | 2m 31s       | 37.75%           |

| 7552             | 2m 33s       | 38%              |

| 7616             | 2m 34s       | 38.5%            |

| 7680             | 2m 35s       | 38.75%           |

| 7744             | 2m 36s       | 39%              |

| 7808             | 2m 38s       | 39.25%           |

| 7872             | 2m 39s       | 39.75%           |

| 7936             | 2m 40s       | 40%              |

| 8000             | 2m 41s       | 40.25%           |

| 8064             | 2m 42s       | 40.75%           |

| 8128             | 2m 44s       | 41%              |

| 8192             | 2m 45s       | 41.25%           |

| 8256             | 2m 46s       | 41.5%            |

| 8320             | 2m 47s       | 42%              |

| 8384             | 2m 49s       | 42.25%           |

| 8448             | 2m 50s       | 42.5%            |

| 8512             | 2m 51s       | 43%              |

| 8576             | 2m 53s       | 43.25%           |

| 8640             | 2m 54s       | 43.5%            |

| 8704             | 2m 55s       | 44%              |

| 8768             | 2m 57s       | 44.25%           |

| 8832             | 2m 58s       | 44.5%            |

| 8896             | 2m 59s       | 44.75%           |

| 8960             | 3m 1s        | 45.25%           |

| 9024             | 3m 2s        | 45.5%            |

| 9088             | 3m 4s        | 45.75%           |

| 9152             | 3m 5s        | 46.25%           |

| 9216             | 3m 6s        | 46.5%            |

| 9280             | 3m 8s        | 46.75%           |

| 9344             | 3m 9s        | 47%              |

| 9408             | 3m 10s       | 47.5%            |

| 9472             | 3m 11s       | 47.75%           |

| 9536             | 3m 13s       | 48%              |

| 9600             | 3m 14s       | 48.5%            |

| 9664             | 3m 15s       | 48.75%           |

| 9728             | 3m 17s       | 49%              |

| 9792             | 3m 18s       | 49.5%            |

| 9856             | 3m 19s       | 49.75%           |

| 9920             | 3m 21s       | 50%              |

| 9984             | 3m 22s       | 50.25%           |

| 10048            | 3m 23s       | 50.75%           |

| 10112            | 3m 25s       | 51%              |

| 10176            | 3m 26s       | 51.25%           |

| 10240            | 3m 27s       | 51.75%           |

| 10304            | 3m 28s       | 52%              |

| 10368            | 3m 30s       | 52.25%           |

| 10432            | 3m 31s       | 52.75%           |

| 10496            | 3m 32s       | 53%              |

| 10560            | 3m 34s       | 53.25%           |

| 10624            | 3m 35s       | 53.5%            |

| 10688            | 3m 36s       | 54%              |

| 10752            | 3m 37s       | 54.25%           |

| 10816            | 3m 39s       | 54.5%            |

| 10880            | 3m 40s       | 55%              |

| 10944            | 3m 41s       | 55.25%           |

| 11008            | 3m 43s       | 55.5%            |

| 11072            | 3m 44s       | 55.75%           |

| 11136            | 3m 45s       | 56.25%           |

| 11200            | 3m 46s       | 56.5%            |

| 11264            | 3m 48s       | 56.75%           |

| 11328            | 3m 49s       | 57.25%           |

| 11392            | 3m 50s       | 57.5%            |

| 11456            | 3m 52s       | 57.75%           |

| 11520            | 3m 53s       | 58.25%           |

| 11584            | 3m 54s       | 58.5%            |

| 11648            | 3m 56s       | 58.75%           |

| 11712            | 3m 57s       | 59%              |

| 11776            | 3m 59s       | 59.5%            |

| 11840            | 4m 0s        | 59.75%           |

| 11904            | 4m 1s        | 60%              |

| 11968            | 4m 2s        | 60.5%            |

| 12032            | 4m 4s        | 60.75%           |

| 12096            | 4m 5s        | 61%              |

| 12160            | 4m 6s        | 61.25%           |

| 12224            | 4m 8s        | 61.75%           |

| 12288            | 4m 9s        | 62%              |

| 12352            | 4m 10s       | 62.25%           |

| 12416            | 4m 12s       | 62.75%           |

| 12480            | 4m 13s       | 63%              |

| 12544            | 4m 14s       | 63.25%           |

| 12608            | 4m 15s       | 63.75%           |

| 12672            | 4m 17s       | 64%              |

| 12736            | 4m 18s       | 64.25%           |

| 12800            | 4m 19s       | 64.5%            |

| 12864            | 4m 21s       | 65%              |

| 12928            | 4m 22s       | 65.25%           |

| 12992            | 4m 23s       | 65.5%            |

| 13056            | 4m 25s       | 66%              |

| 13120            | 4m 26s       | 66.25%           |

| 13184            | 4m 27s       | 66.5%            |

| 13248            | 4m 29s       | 66.75%           |

| 13312            | 4m 30s       | 67.25%           |

| 13376            | 4m 31s       | 67.5%            |

| 13440            | 4m 32s       | 67.75%           |

| 13504            | 4m 34s       | 68.25%           |

| 13568            | 4m 35s       | 68.5%            |

| 13632            | 4m 36s       | 68.75%           |

| 13696            | 4m 38s       | 69.25%           |

| 13760            | 4m 39s       | 69.5%            |

| 13824            | 4m 40s       | 69.75%           |

| 13888            | 4m 42s       | 70%              |

| 13952            | 4m 43s       | 70.5%            |

| 14016            | 4m 44s       | 70.75%           |

| 14080            | 4m 45s       | 71%              |

| 14144            | 4m 47s       | 71.5%            |

| 14208            | 4m 48s       | 71.75%           |

| 14272            | 4m 49s       | 72%              |

| 14336            | 4m 51s       | 72.25%           |

| 14400            | 4m 52s       | 72.75%           |

| 14464            | 4m 53s       | 73%              |

| 14528            | 4m 55s       | 73.25%           |

| 14592            | 4m 56s       | 73.75%           |

| 14656            | 4m 57s       | 74%              |

| 14720            | 4m 58s       | 74.25%           |

| 14784            | 5m 0s        | 74.75%           |

| 14848            | 5m 1s        | 75%              |

| 14912            | 5m 2s        | 75.25%           |

| 14976            | 5m 4s        | 75.5%            |

| 15040            | 5m 5s        | 76%              |

| 15104            | 5m 6s        | 76.25%           |

| 15168            | 5m 7s        | 76.5%            |

| 15232            | 5m 9s        | 77%              |

| 15296            | 5m 10s       | 77.25%           |

| 15360            | 5m 11s       | 77.5%            |

| 15424            | 5m 13s       | 77.75%           |

| 15488            | 5m 14s       | 78.25%           |

| 15552            | 5m 15s       | 78.5%            |

| 15616            | 5m 17s       | 78.75%           |

| 15680            | 5m 18s       | 79.25%           |

| 15744            | 5m 19s       | 79.5%            |

| 15808            | 5m 21s       | 79.75%           |

| 15872            | 5m 22s       | 80.25%           |

| 15936            | 5m 23s       | 80.5%            |

| 16000            | 5m 24s       | 80.75%           |

| 16064            | 5m 26s       | 81%              |

| 16128            | 5m 27s       | 81.5%            |

| 16192            | 5m 28s       | 81.75%           |

| 16256            | 5m 30s       | 82%              |

| 16320            | 5m 31s       | 82.5%            |

| 16384            | 5m 32s       | 82.75%           |

| 16448            | 5m 34s       | 83%              |

| 16512            | 5m 35s       | 83.25%           |

| 16576            | 5m 36s       | 83.75%           |

| 16640            | 5m 37s       | 84%              |

| 16704            | 5m 39s       | 84.25%           |

| 16768            | 5m 40s       | 84.75%           |

| 16832            | 5m 41s       | 85%              |

| 16896            | 5m 43s       | 85.25%           |

| 16960            | 5m 44s       | 85.75%           |

| 17024            | 5m 45s       | 86%              |

| 17088            | 5m 47s       | 86.25%           |

| 17152            | 5m 48s       | 86.5%            |

| 17216            | 5m 49s       | 87%              |

| 17280            | 5m 51s       | 87.25%           |

| 17344            | 5m 52s       | 87.5%            |

| 17408            | 5m 53s       | 88%              |

| 17472            | 5m 54s       | 88.25%           |

| 17536            | 5m 56s       | 88.5%            |

| 17600            | 5m 57s       | 88.75%           |

| 17664            | 5m 58s       | 89.25%           |

| 17728            | 6m 0s        | 89.5%            |

| 17792            | 6m 1s        | 89.75%           |

| 17856            | 6m 2s        | 90.25%           |

| 17920            | 6m 4s        | 90.5%            |

| 17984            | 6m 5s        | 90.75%           |

| 18048            | 6m 6s        | 91.25%           |

| 18112            | 6m 7s        | 91.5%            |

| 18176            | 6m 9s        | 91.75%           |

| 18240            | 6m 10s       | 92%              |

| 18304            | 6m 11s       | 92.5%            |

| 18368            | 6m 13s       | 92.75%           |

| 18432            | 6m 14s       | 93%              |

| 18496            | 6m 15s       | 93.5%            |

| 18560            | 6m 17s       | 93.75%           |

| 18624            | 6m 18s       | 94%              |

| 18688            | 6m 19s       | 94.25%           |

| 18752            | 6m 21s       | 94.75%           |

| 18816            | 6m 22s       | 95%              |

| 18880            | 6m 23s       | 95.25%           |

| 18944            | 6m 24s       | 95.75%           |

| 19008            | 6m 26s       | 96%              |

| 19072            | 6m 27s       | 96.25%           |

| 19328            | 6m 32s       | 97.5%            |

| 19392            | 6m 34s       | 98%              |

| 19456            | 6m 35s       | 98.25%           |

+------------------+--------------+------------------+

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 18760

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 2048

Number of coefficients      : 2049

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 1         | 3        | 0.000053  | 1.013662     | 0.977239          | 0.982090            |

| 2         | 6        | 0.500000  | 2.375037     | 0.980544          | 0.987065            |

| 3         | 7        | 0.500000  | 3.061365     | 0.982463          | 0.988060            |

| 4         | 8        | 0.500000  | 3.720258     | 0.983742          | 0.988060            |

| 5         | 9        | 0.500000  | 4.449610     | 0.984861          | 0.988060            |

| 10        | 14       | 0.500000  | 8.114968     | 0.990991          | 0.994030            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

Completed (Iteration limit reached).

This model may not be optimal. To improve it, consider increasing `max_iterations`.

| 19456            | 6m 35s       | 98.25%           |

Logistic regression

+--------------------------------------------------------

Number of examples          : 18760

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 2048

Number of coefficients      : 2049

Starting L-BFGS

+--------------------------------------------------------

In [6]:
model

Class                                    : ImageClassifier

Schema
------
Number of classes                        : 2
Number of feature columns                : 1
Input image shape                        : (3, 224, 224)

Training summary
----------------
Number of examples                       : 18760
Training loss                            : 1103.7099
Training time (sec)                      : 504.6233

In [7]:
predictions = model.predict(test_data)

In [8]:
predictions

dtype: str
Rows: 5004
['cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'dog', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', 'cat', ... ]

**Evaluate the model and save the results**

In [9]:
metrics = model.evaluate(test_data)

Analyzing and extracting image features.

+------------------+--------------+------------------+

| Images Processed | Elapsed Time | Percent Complete |

+------------------+--------------+------------------+

| 0                | 6.065ms      | 0%               |

| 0                | 6.194ms      | 0%               |

| 0                | 6.309ms      | 0%               |

| 64               | 1.92s        | 1.25%            |

| 128              | 3.16s        | 2.5%             |

| 192              | 4.44s        | 3.75%            |

| 256              | 5.75s        | 5%               |

| 320              | 7.07s        | 6.25%            |

| 384              | 8.39s        | 7.5%             |

| 448              | 9.72s        | 8.75%            |

| 512              | 11.05s       | 10%              |

| 576              | 12.37s       | 11.25%           |

| 640              | 13.73s       | 12.5%            |

| 704              | 14.98s       | 13.75%           |

| 768              | 16.24s       | 15%              |

| 832              | 17.47s       | 16.25%           |

| 896              | 18.78s       | 17.5%            |

| 960              | 20.08s       | 18.75%           |

| 1024             | 21.41s       | 20.25%           |

| 1088             | 22.74s       | 21.5%            |

| 1152             | 24.06s       | 22.75%           |

| 1216             | 25.38s       | 24%              |

| 1280             | 26.71s       | 25.25%           |

| 1344             | 28.04s       | 26.5%            |

| 1408             | 29.29s       | 27.75%           |

| 1472             | 30.53s       | 29%              |

| 1536             | 31.77s       | 30.25%           |

| 1600             | 33.11s       | 31.5%            |

| 1664             | 34.44s       | 32.75%           |

| 1728             | 35.75s       | 34%              |

| 1792             | 37.08s       | 35.25%           |

| 1856             | 38.50s       | 36.5%            |

| 1920             | 39.82s       | 37.75%           |

| 1984             | 41.15s       | 39%              |

| 2048             | 42.43s       | 40.5%            |

| 2112             | 43.67s       | 41.75%           |

| 2176             | 44.93s       | 43%              |

| 2240             | 46.27s       | 44.25%           |

| 2304             | 47.66s       | 45.5%            |

| 2368             | 49.06s       | 46.75%           |

| 2432             | 50.44s       | 48%              |

| 2496             | 51.80s       | 49.25%           |

| 2560             | 53.17s       | 50.5%            |

| 2624             | 54.54s       | 51.75%           |

| 2688             | 55.84s       | 53%              |

| 2752             | 57.08s       | 54.25%           |

| 2816             | 58.31s       | 55.5%            |

| 2880             | 59.60s       | 56.75%           |

| 2944             | 1m 0s        | 58%              |

| 3008             | 1m 2s        | 59.25%           |

| 3072             | 1m 3s        | 60.75%           |

| 3136             | 1m 5s        | 62%              |

| 3200             | 1m 6s        | 63.25%           |

| 3264             | 1m 7s        | 64.5%            |

| 3328             | 1m 9s        | 65.75%           |

| 3392             | 1m 10s       | 67%              |

| 3456             | 1m 11s       | 68.25%           |

| 3520             | 1m 12s       | 69.5%            |

| 3584             | 1m 14s       | 70.75%           |

| 3648             | 1m 15s       | 72%              |

| 3712             | 1m 16s       | 73.25%           |

| 3776             | 1m 18s       | 74.5%            |

| 3840             | 1m 19s       | 75.75%           |

| 3904             | 1m 20s       | 77%              |

| 3968             | 1m 22s       | 78.25%           |

| 4032             | 1m 23s       | 79.5%            |

| 4096             | 1m 24s       | 81%              |

| 4160             | 1m 26s       | 82.25%           |

| 4224             | 1m 27s       | 83.5%            |

| 4352             | 1m 29s       | 86%              |

| 4416             | 1m 31s       | 87.25%           |

| 4480             | 1m 32s       | 88.5%            |

| 4544             | 1m 33s       | 89.75%           |

| 4608             | 1m 35s       | 91%              |

+------------------+--------------+------------------+

| 4608             | 1m 35s       | 91%              |

In [10]:
metrics

{'accuracy': 0.9908073541167066,
 'auc': 0.9981023265771733,
 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |     cat      |       dog       |   22  |
 |     cat      |       cat       |  2512 |
 |     dog      |       cat       |   24  |
 |     dog      |       dog       |  2446 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.9906844876468206,
 'log_loss': 0.05375924860056555,
 'precision': 0.9910858995137763,
 'recall': 0.9902834008097166,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+----------------------+--------------------+------+------+
 | threshold |         fpr          |        tpr         |  p   |  n   |
 +-----------+----------------------+--------------------+------+------+
 |  

In [11]:
print(metrics['accuracy'])

0.9908073541167066


In [12]:
model.save('myCatsDogs.model')

**Export for use in Core ML**

In [13]:
model.export_coreml('myCatsDogs.mlmodel')

**Predictions for some raw data**

In [14]:
new_cats_dogs = tc.image_analysis.load_images('newcatsdogs')

In [15]:
new_cats_dogs

Columns:
	path	str
	image	Image

Rows: 8

Data:
+-------------------------------+------------------------+
|              path             |         image          |
+-------------------------------+------------------------+
| newcatsdogs/download (1).jpeg | Height: 119 Width: 190 |
| newcatsdogs/download (2).jpeg | Height: 121 Width: 119 |
| newcatsdogs/download (3).jpeg | Height: 119 Width: 159 |
| newcatsdogs/download (4).jpeg | Height: 125 Width: 160 |
| newcatsdogs/download (5).jpeg | Height: 125 Width: 167 |
| newcatsdogs/download (6).jpeg | Height: 118 Width: 206 |
| newcatsdogs/download (7).jpeg | Height: 119 Width: 159 |
|   newcatsdogs/download.jpeg   | Height: 121 Width: 182 |
+-------------------------------+------------------------+
[8 rows x 2 columns]

In [16]:
new_cats_dogs['predictions'] = model.predict(new_cats_dogs)

In [17]:
new_cats_dogs['predictions'] 

dtype: str
Rows: 8
['cat', 'cat', 'cat', 'dog', 'dog', 'dog', 'dog', 'cat']

In [18]:
new_cats_dogs.explore()

Materializing SFrame...

Done.

# Advanced Usage

## Pre-trained models

The image classifier toolkit is based on a technique known as transfer learning. 
The advanced options let you select from a set of pre-trained models which can result in a model having various size, performance, and accuracy characteristics.

Like following option, you can change to use [squeezenet](https://apple.github.io/turicreate/docs/api/generated/turicreate.image_classifier.create.html) which can trade off some accuracy for a smaller model with a lower memory and disk foot-print.

In [18]:
model = tc.image_classifier.create(train_data, target = 'label', model = 'squeezenet_v1.1')

Downloading base mlmodel

Downloading https://docs-assets.developer.apple.com/coreml/models/SqueezeNet.mlmodel to /var/folders/m4/8rlv2k612l9g_hd_gmmlbg5m0000gp/T/model_cache/squeezenet_v1.1.mlmodel

Analyzing and extracting image features.

+------------------+--------------+------------------+

| Images Processed | Elapsed Time | Percent Complete |

+------------------+--------------+------------------+

| 0                | 144us        | 0%               |

| 0                | 192us        | 0%               |

| 0                | 221us        | 0%               |

| 64               | 871.439ms    | 0.25%            |

| 128              | 1.09s        | 0.5%             |

| 192              | 1.29s        | 0.75%            |

| 256              | 1.50s        | 1.25%            |

| 320              | 1.70s        | 1.5%             |

| 384              | 1.90s        | 1.75%            |

| 448              | 2.10s        | 2.25%            |

| 512              | 2.30s        | 2.5%             |

| 576              | 2.50s        | 2.75%            |

| 640              | 2.71s        | 3%               |

| 704              | 2.91s        | 3.5%             |

| 768              | 3.12s        | 3.75%            |

| 832              | 3.32s        | 4%               |

| 896              | 3.52s        | 4.5%             |

| 960              | 3.74s        | 4.75%            |

| 1024             | 3.97s        | 5%               |

| 1088             | 4.18s        | 5.5%             |

| 1152             | 4.39s        | 5.75%            |

| 1216             | 4.59s        | 6%               |

| 1280             | 4.80s        | 6.25%            |

| 1344             | 5.00s        | 6.75%            |

| 1408             | 5.21s        | 7%               |

| 1472             | 5.42s        | 7.25%            |

| 1536             | 5.64s        | 7.75%            |

| 1600             | 5.85s        | 8%               |

| 1664             | 6.06s        | 8.25%            |

| 1728             | 6.26s        | 8.5%             |

| 1792             | 6.49s        | 9%               |

| 1856             | 6.69s        | 9.25%            |

| 1920             | 6.90s        | 9.5%             |

| 1984             | 7.10s        | 10%              |

| 2048             | 7.30s        | 10.25%           |

| 2112             | 7.50s        | 10.5%            |

| 2176             | 7.70s        | 11%              |

| 2240             | 7.91s        | 11.25%           |

| 2304             | 8.11s        | 11.5%            |

| 2368             | 8.31s        | 11.75%           |

| 2432             | 8.51s        | 12.25%           |

| 2496             | 8.71s        | 12.5%            |

| 2560             | 8.91s        | 12.75%           |

| 2624             | 9.11s        | 13.25%           |

| 2688             | 9.31s        | 13.5%            |

| 2752             | 9.51s        | 13.75%           |

| 2816             | 9.72s        | 14%              |

| 2880             | 9.92s        | 14.5%            |

| 2944             | 10.13s       | 14.75%           |

| 3008             | 10.32s       | 15%              |

| 3072             | 10.52s       | 15.5%            |

| 3136             | 10.72s       | 15.75%           |

| 3200             | 10.92s       | 16%              |

| 3264             | 11.12s       | 16.5%            |

| 3328             | 11.32s       | 16.75%           |

| 3392             | 11.52s       | 17%              |

| 3456             | 11.71s       | 17.25%           |

| 3520             | 11.91s       | 17.75%           |

| 3584             | 12.10s       | 18%              |

| 3648             | 12.30s       | 18.25%           |

| 3712             | 12.49s       | 18.75%           |

| 3776             | 12.69s       | 19%              |

| 3840             | 12.89s       | 19.25%           |

| 3904             | 13.08s       | 19.5%            |

| 3968             | 13.28s       | 20%              |

| 4032             | 13.47s       | 20.25%           |

| 4096             | 13.67s       | 20.5%            |

| 4160             | 13.86s       | 21%              |

| 4224             | 14.06s       | 21.25%           |

| 4288             | 14.25s       | 21.5%            |

| 4352             | 14.45s       | 22%              |

| 4416             | 14.65s       | 22.25%           |

| 4480             | 14.84s       | 22.5%            |

| 4544             | 15.04s       | 22.75%           |

| 4608             | 15.23s       | 23.25%           |

| 4672             | 15.43s       | 23.5%            |

| 4736             | 15.62s       | 23.75%           |

| 4800             | 15.82s       | 24.25%           |

| 4864             | 16.02s       | 24.5%            |

| 4928             | 16.21s       | 24.75%           |

| 4992             | 16.41s       | 25%              |

| 5056             | 16.60s       | 25.5%            |

| 5120             | 16.80s       | 25.75%           |

| 5184             | 17.00s       | 26%              |

| 5248             | 17.20s       | 26.5%            |

| 5312             | 17.40s       | 26.75%           |

| 5376             | 17.60s       | 27%              |

| 5440             | 17.80s       | 27.5%            |

| 5504             | 18.00s       | 27.75%           |

| 5568             | 18.20s       | 28%              |

| 5632             | 18.40s       | 28.25%           |

| 5696             | 18.60s       | 28.75%           |

| 5760             | 18.80s       | 29%              |

| 5824             | 19.00s       | 29.25%           |

| 5888             | 19.20s       | 29.75%           |

| 5952             | 19.40s       | 30%              |

| 6016             | 19.60s       | 30.25%           |

| 6080             | 19.80s       | 30.5%            |

| 6144             | 20.00s       | 31%              |

| 6208             | 20.20s       | 31.25%           |

| 6272             | 20.39s       | 31.5%            |

| 6336             | 20.59s       | 32%              |

| 6400             | 20.79s       | 32.25%           |

| 6464             | 20.99s       | 32.5%            |

| 6528             | 21.18s       | 33%              |

| 6592             | 21.38s       | 33.25%           |

| 6656             | 21.58s       | 33.5%            |

| 6720             | 21.78s       | 33.75%           |

| 6784             | 21.97s       | 34.25%           |

| 6848             | 22.17s       | 34.5%            |

| 6912             | 22.37s       | 34.75%           |

| 6976             | 22.56s       | 35.25%           |

| 7040             | 22.76s       | 35.5%            |

| 7104             | 22.95s       | 35.75%           |

| 7168             | 23.15s       | 36%              |

| 7232             | 23.34s       | 36.5%            |

| 7296             | 23.54s       | 36.75%           |

| 7360             | 23.73s       | 37%              |

| 7424             | 23.93s       | 37.5%            |

| 7488             | 24.12s       | 37.75%           |

| 7552             | 24.32s       | 38%              |

| 7616             | 24.52s       | 38.5%            |

| 7680             | 24.71s       | 38.75%           |

| 7744             | 24.91s       | 39%              |

| 7808             | 25.10s       | 39.25%           |

| 7872             | 25.30s       | 39.75%           |

| 7936             | 25.49s       | 40%              |

| 8000             | 25.69s       | 40.25%           |

| 8064             | 25.88s       | 40.75%           |

| 8128             | 26.08s       | 41%              |

| 8192             | 26.27s       | 41.25%           |

| 8256             | 26.47s       | 41.5%            |

| 8320             | 26.66s       | 42%              |

| 8384             | 26.86s       | 42.25%           |

| 8448             | 27.05s       | 42.5%            |

| 8512             | 27.25s       | 43%              |

| 8576             | 27.44s       | 43.25%           |

| 8640             | 27.64s       | 43.5%            |

| 8704             | 27.84s       | 44%              |

| 8768             | 28.03s       | 44.25%           |

| 8832             | 28.23s       | 44.5%            |

| 8896             | 28.42s       | 44.75%           |

| 8960             | 28.62s       | 45.25%           |

| 9024             | 28.82s       | 45.5%            |

| 9088             | 29.01s       | 45.75%           |

| 9152             | 29.21s       | 46.25%           |

| 9216             | 29.40s       | 46.5%            |

| 9280             | 29.60s       | 46.75%           |

| 9344             | 29.79s       | 47%              |

| 9408             | 29.99s       | 47.5%            |

| 9472             | 30.18s       | 47.75%           |

| 9536             | 30.38s       | 48%              |

| 9600             | 30.58s       | 48.5%            |

| 9664             | 30.78s       | 48.75%           |

| 9728             | 30.99s       | 49%              |

| 9792             | 31.19s       | 49.5%            |

| 9856             | 31.39s       | 49.75%           |

| 9920             | 31.59s       | 50%              |

| 9984             | 31.79s       | 50.25%           |

| 10048            | 31.99s       | 50.75%           |

| 10112            | 32.19s       | 51%              |

| 10176            | 32.40s       | 51.25%           |

| 10240            | 32.60s       | 51.75%           |

| 10304            | 32.84s       | 52%              |

| 10368            | 33.05s       | 52.25%           |

| 10432            | 33.28s       | 52.75%           |

| 10496            | 33.51s       | 53%              |

| 10560            | 33.74s       | 53.25%           |

| 10624            | 33.95s       | 53.5%            |

| 10688            | 34.15s       | 54%              |

| 10752            | 34.38s       | 54.25%           |

| 10816            | 34.59s       | 54.5%            |

| 10880            | 34.80s       | 55%              |

| 10944            | 35.00s       | 55.25%           |

| 11008            | 35.22s       | 55.5%            |

| 11072            | 35.43s       | 55.75%           |

| 11136            | 35.62s       | 56.25%           |

| 11200            | 35.84s       | 56.5%            |

| 11264            | 36.04s       | 56.75%           |

| 11328            | 36.23s       | 57.25%           |

| 11392            | 36.43s       | 57.5%            |

| 11456            | 36.66s       | 57.75%           |

| 11520            | 36.86s       | 58.25%           |

| 11584            | 37.08s       | 58.5%            |

| 11648            | 37.29s       | 58.75%           |

| 11712            | 37.49s       | 59%              |

| 11776            | 37.70s       | 59.5%            |

| 11840            | 37.91s       | 59.75%           |

| 11904            | 38.12s       | 60%              |

| 11968            | 38.32s       | 60.5%            |

| 12032            | 38.55s       | 60.75%           |

| 12096            | 38.78s       | 61%              |

| 12160            | 39.00s       | 61.25%           |

| 12224            | 39.21s       | 61.75%           |

| 12288            | 39.41s       | 62%              |

| 12352            | 39.62s       | 62.25%           |

| 12416            | 39.83s       | 62.75%           |

| 12480            | 40.05s       | 63%              |

| 12544            | 40.28s       | 63.25%           |

| 12608            | 40.51s       | 63.75%           |

| 12672            | 40.74s       | 64%              |

| 12736            | 40.97s       | 64.25%           |

| 12800            | 41.20s       | 64.5%            |

| 12864            | 41.43s       | 65%              |

| 12928            | 41.66s       | 65.25%           |

| 12992            | 41.88s       | 65.5%            |

| 13056            | 42.08s       | 66%              |

| 13120            | 42.31s       | 66.25%           |

| 13184            | 42.53s       | 66.5%            |

| 13248            | 42.75s       | 66.75%           |

| 13312            | 42.96s       | 67.25%           |

| 13376            | 43.19s       | 67.5%            |

| 13440            | 43.40s       | 67.75%           |

| 13504            | 43.61s       | 68.25%           |

| 13568            | 43.81s       | 68.5%            |

| 13632            | 44.01s       | 68.75%           |

| 13696            | 44.22s       | 69.25%           |

| 13760            | 44.42s       | 69.5%            |

| 13824            | 44.62s       | 69.75%           |

| 13888            | 44.82s       | 70%              |

| 13952            | 45.02s       | 70.5%            |

| 14016            | 45.23s       | 70.75%           |

| 14080            | 45.43s       | 71%              |

| 14144            | 45.64s       | 71.5%            |

| 14208            | 45.85s       | 71.75%           |

| 14272            | 46.06s       | 72%              |

| 14336            | 46.26s       | 72.25%           |

| 14400            | 46.47s       | 72.75%           |

| 14464            | 46.68s       | 73%              |

| 14528            | 46.90s       | 73.25%           |

| 14592            | 47.11s       | 73.75%           |

| 14656            | 47.32s       | 74%              |

| 14720            | 47.53s       | 74.25%           |

| 14784            | 47.74s       | 74.75%           |

| 14848            | 47.95s       | 75%              |

| 14912            | 48.16s       | 75.25%           |

| 14976            | 48.37s       | 75.5%            |

| 15040            | 48.58s       | 76%              |

| 15104            | 48.78s       | 76.25%           |

| 15168            | 48.99s       | 76.5%            |

| 15232            | 49.20s       | 77%              |

| 15296            | 49.41s       | 77.25%           |

| 15360            | 49.62s       | 77.5%            |

| 15424            | 49.82s       | 77.75%           |

| 15488            | 50.03s       | 78.25%           |

| 15552            | 50.23s       | 78.5%            |

| 15616            | 50.43s       | 78.75%           |

| 15680            | 50.63s       | 79.25%           |

| 15744            | 50.84s       | 79.5%            |

| 15808            | 51.04s       | 79.75%           |

| 15872            | 51.24s       | 80.25%           |

| 15936            | 51.45s       | 80.5%            |

| 16000            | 51.65s       | 80.75%           |

| 16064            | 51.85s       | 81%              |

| 16128            | 52.06s       | 81.5%            |

| 16192            | 52.26s       | 81.75%           |

| 16256            | 52.46s       | 82%              |

| 16320            | 52.66s       | 82.5%            |

| 16384            | 52.87s       | 82.75%           |

| 16448            | 53.07s       | 83%              |

| 16512            | 53.27s       | 83.25%           |

| 16576            | 53.47s       | 83.75%           |

| 16640            | 53.67s       | 84%              |

| 16704            | 53.87s       | 84.25%           |

| 16768            | 54.08s       | 84.75%           |

| 16832            | 54.28s       | 85%              |

| 16896            | 54.48s       | 85.25%           |

| 16960            | 54.68s       | 85.75%           |

| 17024            | 54.88s       | 86%              |

| 17088            | 55.11s       | 86.25%           |

| 17152            | 55.33s       | 86.5%            |

| 17216            | 55.56s       | 87%              |

| 17280            | 55.78s       | 87.25%           |

| 17344            | 56.01s       | 87.5%            |

| 17408            | 56.23s       | 88%              |

| 17472            | 56.43s       | 88.25%           |

| 17536            | 56.64s       | 88.5%            |

| 17600            | 56.87s       | 88.75%           |

| 17664            | 57.10s       | 89.25%           |

| 17728            | 57.33s       | 89.5%            |

| 17792            | 57.55s       | 89.75%           |

| 17856            | 57.78s       | 90.25%           |

| 17920            | 58.01s       | 90.5%            |

| 17984            | 58.23s       | 90.75%           |

| 18048            | 58.45s       | 91.25%           |

| 18112            | 58.68s       | 91.5%            |

| 18176            | 58.90s       | 91.75%           |

| 18240            | 59.12s       | 92%              |

| 18304            | 59.32s       | 92.5%            |

| 18368            | 59.52s       | 92.75%           |

| 18432            | 59.73s       | 93%              |

| 18496            | 59.93s       | 93.5%            |

| 18560            | 1m 0s        | 93.75%           |

| 18624            | 1m 0s        | 94%              |

| 18688            | 1m 0s        | 94.25%           |

| 18752            | 1m 0s        | 94.75%           |

| 18816            | 1m 0s        | 95%              |

| 18880            | 1m 1s        | 95.25%           |

| 18944            | 1m 1s        | 95.75%           |

| 19008            | 1m 1s        | 96%              |

| 19072            | 1m 1s        | 96.25%           |

| 19328            | 1m 2s        | 97.5%            |

| 19392            | 1m 3s        | 98%              |

| 19456            | 1m 3s        | 98.25%           |

+------------------+--------------+------------------+

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 18835

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 1000

Number of coefficients      : 1001

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 1         | 6        | 0.000005  | 1.517308     | 0.529652          | 0.562366            |

| 2         | 8        | 1.000000  | 2.292306     | 0.828829          | 0.839785            |

| 3         | 9        | 1.000000  | 2.809328     | 0.907778          | 0.920430            |

| 4         | 10       | 1.000000  | 3.301846     | 0.907619          | 0.918280            |

| 5         | 11       | 1.000000  | 3.783262     | 0.501088          | 0.465591            |

| 10        | 23       | 1.000000  | 7.833680     | 0.915530          | 0.920430            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

Completed (Iteration limit reached).

This model may not be optimal. To improve it, consider increasing `max_iterations`.

| 19456            | 1m 3s        | 98.25%           |

Logistic regression:

+--------------------------------------------------------

Number of examples          : 18835

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 1000

Number of coefficients      : 1001

Starting L-BFGS

+--------------------------------------------------------

In [19]:
 predictions = model.predict(test_data)

In [20]:
metrics = model.evaluate(test_data)

Analyzing and extracting image features.

+------------------+--------------+------------------+

| Images Processed | Elapsed Time | Percent Complete |

+------------------+--------------+------------------+

| 0                | 5.657ms      | 0%               |

| 0                | 5.695ms      | 0%               |

| 0                | 5.753ms      | 0%               |

| 64               | 931.507ms    | 1.25%            |

| 128              | 1.16s        | 2.5%             |

| 192              | 1.39s        | 3.75%            |

| 256              | 1.61s        | 5%               |

| 320              | 1.81s        | 6.25%            |

| 384              | 2.02s        | 7.5%             |

| 448              | 2.23s        | 8.75%            |

| 512              | 2.44s        | 10%              |

| 576              | 2.64s        | 11.25%           |

| 640              | 2.84s        | 12.5%            |

| 704              | 3.05s        | 13.75%           |

| 768              | 3.27s        | 15%              |

| 832              | 3.49s        | 16.25%           |

| 896              | 3.71s        | 17.5%            |

| 960              | 3.93s        | 18.75%           |

| 1024             | 4.14s        | 20.25%           |

| 1088             | 4.34s        | 21.5%            |

| 1152             | 4.55s        | 22.75%           |

| 1216             | 4.77s        | 24%              |

| 1280             | 4.98s        | 25.25%           |

| 1344             | 5.18s        | 26.5%            |

| 1408             | 5.38s        | 27.75%           |

| 1472             | 5.59s        | 29%              |

| 1536             | 5.79s        | 30.25%           |

| 1600             | 6.00s        | 31.5%            |

| 1664             | 6.21s        | 32.75%           |

| 1728             | 6.41s        | 34%              |

| 1792             | 6.62s        | 35.25%           |

| 1856             | 6.84s        | 36.5%            |

| 1920             | 7.04s        | 37.75%           |

| 1984             | 7.25s        | 39%              |

| 2048             | 7.45s        | 40.5%            |

| 2112             | 7.67s        | 41.75%           |

| 2176             | 7.88s        | 43%              |

| 2240             | 8.09s        | 44.25%           |

| 2304             | 8.29s        | 45.5%            |

| 2368             | 8.52s        | 46.75%           |

| 2432             | 8.72s        | 48%              |

| 2496             | 8.92s        | 49.25%           |

| 2560             | 9.12s        | 50.5%            |

| 2624             | 9.32s        | 51.75%           |

| 2688             | 9.52s        | 53%              |

| 2752             | 9.72s        | 54.25%           |

| 2816             | 9.95s        | 55.5%            |

| 2880             | 10.15s       | 56.75%           |

| 2944             | 10.35s       | 58%              |

| 3008             | 10.56s       | 59.25%           |

| 3072             | 10.76s       | 60.75%           |

| 3136             | 10.97s       | 62%              |

| 3200             | 11.17s       | 63.25%           |

| 3264             | 11.37s       | 64.5%            |

| 3328             | 11.57s       | 65.75%           |

| 3392             | 11.77s       | 67%              |

| 3456             | 11.96s       | 68.25%           |

| 3520             | 12.16s       | 69.5%            |

| 3584             | 12.35s       | 70.75%           |

| 3648             | 12.55s       | 72%              |

| 3712             | 12.75s       | 73.25%           |

| 3776             | 12.94s       | 74.5%            |

| 3840             | 13.14s       | 75.75%           |

| 3904             | 13.33s       | 77%              |

| 3968             | 13.53s       | 78.25%           |

| 4032             | 13.72s       | 79.5%            |

| 4096             | 13.92s       | 81%              |

| 4160             | 14.12s       | 82.25%           |

| 4224             | 14.33s       | 83.5%            |

| 4288             | 14.53s       | 84.75%           |

| 4352             | 14.73s       | 86%              |

| 4416             | 14.93s       | 87.25%           |

| 4544             | 15.34s       | 89.75%           |

| 4672             | 15.62s       | 92.25%           |

+------------------+--------------+------------------+

| 4672             | 15.62s       | 92.25%           |

In [25]:
model.save('myCatsDogs_squeezenet.model')

In [26]:
model.export_coreml('myCatsDogs_squeezenet.mlmodel')

In [23]:
new_cats_dogs['predictions'] = model.predict(new_cats_dogs)

In [24]:
new_cats_dogs.explore()

Materializing SFrame

## Using GPUs

GPUs can make creating an image classifier model much faster. If you have macOS 10.13 or higher, Turi Create will automatically use the GPU. If your Linux machine has an NVIDIA GPU, you can setup Turi Create to use the GPU, [see instructions](https://github.com/apple/turicreate/blob/master/LinuxGPU.md).

The turicreate.config.set_num_gpus function allows you to control if GPUs are used:

```
# Use all GPUs (default)
turicreate.config.set_num_gpus(-1)

# Use only 1 GPU
turicreate.config.set_num_gpus(1)

# Use CPU
turicreate.config.set_num_gpus(0)

```